Reworked findContours to reduce C-API usage

mshabunin opened the pull request:
What is done:
* rewritten `findContours` and `icvApproximateChainTC89` using C++ data structures
* extracted LINK_RUNS mode to separate new public functions - `findContoursLinkRuns` (it uses completely different algorithm)
* ~added new public `cv::approximateChainTC89`~ - **:x: decided to hide it**
* enabled chain code output (method = 0, no public enum value for this in C++ yet)
* kept old function as `findContours_old` (exported, but not exposed to user)
* added more tests for findContours (`test_contours_new.cpp`), some tests compare results of old function with new one. Following tests have been added:
  * contours of random rectangle
  * contours of many small (1-2px) blobs
  * contours of random noise
  * backport of old accuracy test
  * separate test for LINK RUNS variant

What is left to be done (can be done now or later):
* improve tests: 
  * some tests have limited verification (e.g. only verify contour sizes)
  * perhaps reference data can be collected and stored
  * maybe more test variants can be added (?)
* add enum value for chain code output and a method of returning starting points (e.g. first 8 elements of returned `vector<uchar>` can represent 2 int point coordinates)
* add documentation for new functions - **:heavy_check_mark: DONE**
* check and improve performance (my experiment showed 0.7x-1.1x some time ago)
* remove old functions completely (?)
* change contour return order (BFS) or allow to select it (?)
* return result tree as-is (?) (new data structures should be exposed, bindings should adapt)

asmorkalov commented on modules/imgproc/src/approx.cpp:
```diff
@@ -103,9 +103,9 @@ CvSeq* icvApproximateChainTC89( CvChain* chain, int header_size,
         s = abs_diff[reader.code - prev_code + 7];
 
-        if( method <= cv::CHAIN_APPROX_SIMPLE )
+        if( method <= CV_CHAIN_APPROX_SIMPLE )
```
Why do you revert C++ const to old C const?

mshabunin replied:
Reverted to older version by mistake, I'll fix it.

asmorkalov commented on modules/imgproc/src/contour_approx.cpp:
```diff
@@ -0,0 +1,351 @@
+        double temp_num = dx1 * dx2 + dy1 * dy2;
+        temp_num =
+            (float) (temp_num /
+                     sqrt( ((double)dx1 * dx1 + (double)dy1 * dy1) *
```
Is NaN possible here? E.g. division by zero or sqrt for negative value?

mshabunin replied:
Should not be possible, there is a check for 0 (dx, dy are ints):
```.cpp
if( (dx1 | dy1) == 0 || (dx2 | dy2) == 0 )
            break;
```
And sum of squares can not be less than zero too, so sqrt should is also safe.

asmorkalov commented on modules/imgproc/include/opencv2/imgproc.hpp:
```diff
@@ -451,6 +451,9 @@ enum ContourApproximationModes {
 };
 
+CV_EXPORTS
+std::vector<Point> approximateChainTC89(std::vector<schar> chain_, Point origin, const int method);
```
Why not CV_EXPORTS_W? What about documentation and reference to paper?

mshabunin replied:
I exported this function to be able to test it (though I haven't make such tests yet), I'm not sure whether is it useful. It transforms chain code to vector of points using one of methods, but there are no other interfaces with chain codes. New findContours can return chain code, but without origin point it is not very useful.

So export of this function is still in question. I wanted to get other opinions. Should we complete chain code output in `findContours` and expose `approximateChain` or hide these functionality?

asmorkalov commented on modules/imgproc/src/contours_common.cpp:
The file is empty. Do we need it?

mshabunin replied:
Yes, I've moved `contourTreeToResults` function there (it was original intention).

asmorkalov commented on modules/imgproc/src/contours_common.hpp:
```diff
@@ -0,0 +1,269 @@
+//                ;
+        return levels_s.empty();
+    }
+//    const TreeNode<T> & getNext()
+//    {
+//        const int idx = levels.front();
+//        CV_Assert(idx >= 0);
+//        levels.pop();
+//        const TreeNode<T> & res = tree.elem(idx);
+//        int cur = res.first_child;
+//        while (cur != -1)
+//        {
+//            levels.push(cur);
+//            cur = tree.elem(cur).next;
+//        }
+//        return res;
+//    }
```
Please remove unused methods.

asmorkalov commented on modules/imgproc/include/opencv2/imgproc.hpp:
```diff
@@ -4031,6 +4034,14 @@ CV_EXPORTS_W void findContours( InputArray image, OutputArrayOfArrays contours,
 CV_EXPORTS void findContours( InputArray image, OutputArrayOfArrays contours,
                               int mode, int method, Point offset = Point());
 
+#ifdef __OPENCV_BUILD
+CV_EXPORTS void findContours_old( InputArray _image, OutputArrayOfArrays _contours,
+                                  OutputArray _hierarchy, int mode, int method,
+                                  Point offset = Point() );
+CV_EXPORTS void findContours_old( InputArray image, OutputArrayOfArrays contours,
+                                  int mode, int method, Point offset = Point());
+#endif
```
I propose to move it to `opencv2/imgproc/imgproc_c.h` or some `details` header. In case, if we need to presume the implementation, it's less painful for 4.x->5.x merges.
What about `findContoursLegacy` ?

mshabunin replied:
Moved to `imgproc/detail/legacy.hpp`, renamed to `findContours_legacy`.

asmorkalov commented on modules/imgproc/src/contour_approx.cpp:
```diff
@@ -0,0 +1,351 @@
+
+
+
+vector<Point> cv::approximateChainTC89(vector<schar> chain, Point origin, const int method)
```
`const Point&`

asmorkalov commented on modules/imgproc/src/contour_approx.cpp:
```diff
@@ -0,0 +1,351 @@
+static vector<ApproxItem> pass_0(const vector<schar> & chain, Point origin, bool isApprox, bool isFull)
+{
+    CV_DbgAssert(chain.size() < (size_t)numeric_limits<int>::max());
+    Point pt = origin;
```
origin is already passed by value. The copy is not needed.

asmorkalov commented on modules/imgproc/src/contour_approx.cpp:
```diff
@@ -0,0 +1,351 @@
+    CV_DbgAssert(chain.size() < (size_t)numeric_limits<int>::max());
+    Point pt = origin;
+    vector<ApproxItem> res;
+    const int len = (int)chain.size();
+    for (int i = 0; i < len; ++i)
```
it's easier to use size_t for both `len` and `i`.

asmorkalov commented on modules/imgproc/include/opencv2/imgproc.hpp:
```diff
@@ -4041,6 +4052,9 @@ A program using pyramid scaling, Canny, contours and contour simplification to f
 squares in the input image.
 */
 
+CV_EXPORTS_W void findContoursLinkRuns(InputArray image, OutputArrayOfArrays contours, OutputArray hierarchy);
+CV_EXPORTS_W void findContoursLinkRuns(InputArray image, OutputArrayOfArrays contours);
```
Could you add documentation for the functions? Do we really need it in public API?

mshabunin replied:
It was separate mode in `findContours` - CV_LINK_RUNS, but it is very different from other modes (Suzuki algorithm), for example it doesn't require internal image copying and threshold. It also supports only RETR_LIST results output. So I thought that it would be better to extract it to separate function. I don't know if it is useful for anyone or have any bugs, because it was not tested or called anywhere in samples. I've added result verification by comparing it with old implementation output in the `link_runs` test, but I haven't checked it beyond that.

I also can not find paper describing this algorithm, maybe @vpisarev could help?

I can remove this function if we decide we don't need it.

asmorkalov commented on modules/imgproc/src/contours_approx.cpp:
```diff
@@ -0,0 +1,353 @@
+        /* calc indices */
+        const int i1 = (i >= k) ? (i - k) : (len - k + i);
+        const int i2 = (i + k < len) ? (i + k) : (i + k - len);
+
+
```
Extra line.

asmorkalov commented on modules/imgproc/src/contours_approx.cpp:
```diff
@@ -0,0 +1,353 @@
+    /* determine support region */
+    for( k = 1;; k++ )
+    {
+        CV_Assert( k <= len );
```
What if add the check to for loop?

mshabunin replied:
This was in the original code and, since this function isn't covered by tests, I tried to preserve this original behavior. As I understand, `k` should't exceed `len` (even `len/2` maybe) and if this happens then it means that input array was not valid, so exception might be more appropriate than continuing.

asmorkalov commented on modules/imgproc/src/contours_approx.cpp:
```diff
@@ -0,0 +1,353 @@
+
+        // Pass 1.
+        // Determines support region for all the remained points */
+        for (int i = 0; i < (int)ares.size(); ++i)
```
looks like if you replace indexes from int to `size_t` then all casts to int can be dropped. It's related to `pass_cleanup`, `clear_until`, `calc_dominance`, etc.

mshabunin replied:
Originally I tried to use `size_t`, but in some places arithmetics requires switching to signed type or comparing with signed values and VS doesn't like it. Also, perhaps 32-bit might be slightly faster or consume less memory? I've added debug assertion to check that main array size is less than max int for safety.

mshabunin replied:
Made these changes too.

asmorkalov commented on modules/imgproc/src/contours_common.cpp:
```diff
@@ -0,0 +1,75 @@
+            if (elem.self() == 0)
+                continue;
+            Vec4i & h_vec = h_mat.at<Vec4i>(i);
+            h_vec = Vec4i(index_mapping.at(elem.next),
+                              index_mapping.at(elem.prev),
+                              index_mapping.at(elem.first_child),
+                              index_mapping.at(elem.parent));
```
formatting.

mshabunin replied:
Can I use this style (https://github.com/opencv/opencv/wiki/.clang_format) for formatting for all new files?

asmorkalov commented on modules/imgproc/src/contours_link.cpp:
```diff
@@ -0,0 +1,405 @@
+
+        cur_point.x = j;
+
+        rns.push_back(LRP(cur_point));
```
Looks like rns has predictable size. It makes sense to use reserve and emplace_back.

mshabunin replied:
`rns` size is not predictable, it depends on number of contours on image and their shape. We can reserve some amount assuming there are at least some contours though.

asmorkalov commented on modules/imgproc/src/contours_new.cpp:
```diff
@@ -0,0 +1,666 @@
+                if( Trait<T>::checkValue(i4, i0) )
+                    break;
+            }
+
+
```
Extra line.

asmorkalov commented on modules/imgproc/src/contours_new.cpp:
```diff
@@ -0,0 +1,666 @@
+
+template <typename T>
+static void icvFetchContourEx(Mat image,
+                              Point start,
```
const &

asmorkalov commented on modules/imgproc/src/contours_new.cpp:
```diff
@@ -0,0 +1,666 @@
+    scanner->nbd = 2;
+    CNode & root = scanner->tree.newElem();
+    CV_Assert(root.self() == 0);
+    root.body.isHole = true;
+    root.body.isHole = true;
```
Duplicate.

mshabunin commented:
@asmorkalov , I've addressed some of the comments, but still not sure what to  do with others:
* can I use clang-format from our wiki for consistent formatting?
* do we choose to export `contourApprox` function and chain code output for `findContours` or hide them?

AleksandrPanov commented:
@mshabunin 
1. could you add info which paper describes `findContoursLinkRuns()` (if you know)?
2. the performance of `findContours()` is very important, this function used in findChessboardCorners, SimpleBlobDetector, findCircleGrid, ArucoDetector, CCheckerDetector, ... and `findContours()` may take up to 50% of the execution time in some cases. Do you have any simple suggestions on how to improve performance?

mshabunin commented:
1. I haven't find the paper for link runs approach, all hopes on @vpisarev  :slightly_smiling_face: 
2. I did some basic optimizations by reserving internal vector sizes with assumption that contours are not very small. Further optimization can be done by using single large vector for all points and only storing ranges in it in each contour record. Additionally we can think about more efficient public interface, e.g. `findContours` could return an object of some new type which would allow users to iterate over contours and access data in each of them.

opencv-alalek commented on modules/imgproc/src/contours_new.cpp:
```diff
@@ -27,69 +27,77 @@ struct Trait
+static const int MASK_NEW = 0x40000000; // 010..000
 static const int MASK_FLAGS = 0xC0000000; // right + new
-static const int MASK_VAL   = 0x3FFFFFFF; // ~flags - pixel label
+static const int MASK_VAL = 0x3FFFFFFF; // ~flags - pixel label
```
Is there clang-format option 'do not touch'?

mshabunin replied:
We can enclosure a block of code between `// clang-format on` - `// clang-format off` comments to ignore it.

But I believe in general it is better to avoid alignment of values or comments in such blocks to simplify future merges. Thus I chose to disable these rules: `AlignConsecutiveAssignments: None` and other `AlignConsecutive*` options (see https://releases.llvm.org/12.0.0/tools/clang/docs/ClangFormatStyleOptions.html for details)

opencv-alalek commented on modules/imgproc/src/contours_new.cpp:
```diff
@@ -271,59 +277,60 @@ static void icvFetchContourEx(Mat &image,
-             || mode == RETR_TREE
-             || mode == RETR_FLOODFILL,
+             mode == RETR_EXTERNAL || mode == RETR_LIST || mode == RETR_CCOMP || mode == RETR_TREE
+                 || mode == RETR_FLOODFILL,
```
less readable

mshabunin replied:
I agree. Unfortunately this is due to some weird clang-format behavior - it doesn't align arguments in a function call as in other places and operator indentation doesn't work too.

It can be fixed in following ways:
1. Add comments after each each piece, then it will only add a single offset (`ContinuationIndentWidth`) - not very good, because it incentivise addition of meaningless comments
2. Move this condition to separate variable - it would be good in general, but not in this case, because of `CV_Check` specifics - we want full condition to be visible in logs
3. Mark `CV_Check` macro as whitespace-sensitive (`WhitespaceSensitiveMacros` option), then it won't touch its internals at all - not good in general as it disables auto-formatting

I decided in this case to leave this specific check call less readable, and only partially fix the next check call, because comment is appropriate there:
```.cpp
    CV_Check(mode,
             mode == RETR_EXTERNAL || mode == RETR_LIST || mode == RETR_CCOMP || mode == RETR_TREE
                 || mode == RETR_FLOODFILL,
             "Wrong extraction mode");

    CV_Check(method,
             method == 0  // raw chain code - no enum value yet
                 || method == CHAIN_APPROX_NONE || method == CHAIN_APPROX_SIMPLE
                 || method == CHAIN_APPROX_TC89_L1 || method == CHAIN_APPROX_TC89_KCOS,
             "Wrong approximation method");
```

Perhaps it would be good to add special macro `CV_CheckIsOneOf(method, 0, CHAIN_APPROX_NONE, ..., "some message")` or some other convenient tool for similar cases.

opencv-alalek commented on modules/imgproc/src/contours_new.cpp:
```diff
@@ -271,59 +277,60 @@ static void icvFetchContourEx(Mat &image,
+    Point lnbd; /* position of the last met contour */
+    schar nbd; /* current mark val */
+    int approx_method1; /* approx method when tracing */
+    int approx_method2; /* final approx method */
```
`do not touch` is preferable here too.

2 or more spaces before comment should help with reading code.

mshabunin replied:
This would require setting option `SpacesBeforeTrailingComments: 2` **and** changing comments to `//`- I'll apply these changes.

opencv-alalek commented on modules/imgproc/src/contours_new.cpp:
```diff
@@ -510,36 +503,33 @@ bool ContourScanner_::contourScan(const int prev, int &p, Point &last_pos, const
-            ((last_pos.y - cur_elem.body.brect.y) < cur_elem.body.brect.height))
+        CNode& cur_elem = tree.elem(cur);
+        if (((last_pos.x - cur_elem.body.brect.x) < cur_elem.body.brect.width)
+            && ((last_pos.y - cur_elem.body.brect.y) < cur_elem.body.brect.height))
```
previous variable is more readable.

mshabunin replied:
In this case I decided to move this condition to separate variable, then it would be aligned according to `BreakBeforeBinaryOperators` and `AlignOperands` rules:
```.cpp
        const bool isBefore = ((last_pos.x - cur_elem.body.brect.x) < cur_elem.body.brect.width)
                           && ((last_pos.y - cur_elem.body.brect.y) < cur_elem.body.brect.height);
        if (isBefore)
        {
```

opencv-alalek commented on modules/imgproc/src/contours_new.cpp:
```diff
@@ -627,37 +619,47 @@ bool ContourScanner_::findNext()
-        || (_contours.kind() == _InputArray::STD_VECTOR_MAT)
-        || (_contours.kind() == _InputArray::STD_VECTOR_UMAT));
+              || (_contours.kind() == _InputArray::STD_VECTOR_MAT)
+              || (_contours.kind() == _InputArray::STD_VECTOR_UMAT));
```
`_contours.kind()` should be aligned to have readable code.

mshabunin replied:
Here we are seeing same situation as in previous comments. We can move this expression outside, but macro would loose its meaning. Or we can add this macro to whitespace-preserve list. I'd prefer to create new check macro to cover these cases.

opencv-alalek commented on modules/imgproc/test/test_contours_new.cpp:
```diff
@@ -89,22 +92,20 @@ TEST_P(Imgproc_FindContours_Modes1, rectangle)
+        const vector<Point> ext_ref {r.tl(),
+                                     r.tl() + Point(0, r.height - 1),
+                                     r.br() + Point(-1, -1),
+                                     r.tl() + Point(r.width - 1, 0)};
```
Previous variant is better.

mshabunin replied:
Seems to be improved by setting `PenaltyIndentedWhitespace: 1`

opencv-alalek commented on modules/imgproc/test/test_contours_new.cpp:
```diff
@@ -446,10 +447,8 @@ TEST_P(Imgproc_FindContours_Modes2, new_accuracy)
-                    || img.at<uchar>(y, x + 1) == 0
-                    || img.at<uchar>(y, x - 1) == 0))
+                && (img.at<uchar>(y - 1, x) == 0 || img.at<uchar>(y + 1, x) == 0
+                    || img.at<uchar>(y, x + 1) == 0 || img.at<uchar>(y, x - 1) == 0))
```
the same, align `.at()` calls one per line

mshabunin replied:
Can not make clang-format to align continuations properly when breaks are before binary operators, perhaps it is some sort of a bug. So I decided to switch to `BreakBeforeBinaryOperators: None` in the end :slightly_frowning_face: This code will be:
```.cpp
            if (img.at<uchar>(y, x) != 0 &&
                ((img.at<uchar>(y - 1, x) == 0) || (img.at<uchar>(y + 1, x) == 0) ||
                 (img.at<uchar>(y, x + 1) == 0) || (img.at<uchar>(y, x - 1) == 0)))
            {
```

opencv-alalek commented on modules/imgproc/test/test_contours_new.cpp:
```diff
@@ -550,29 +549,29 @@ TEST_P(Imgproc_FindContours_Modes2, approx)
+    //    const int r { 20 };
+    //    const Point cut {r, r};
+    //    circle(img, center, r, Scalar(255), FILLED);
+    //    rectangle(img, center, center + cut, Scalar(0), FILLED);
```
`#if 0` is preferable

mshabunin replied:
Will remove this code

opencv-alalek commented on modules/imgproc/test/test_contours_new.cpp:
```diff
@@ -585,13 +584,13 @@ TEST(Imgproc_FindContours, link_runs)
+    //    drawContours(res, contours);
+    //    imshow("res", res);
+    //    imshow("img", img);
+    //    waitKey(0);
```
`cvtest::debugLevel >= 10`
- 1 print/write result data/images
- 10 show windows

mshabunin commented on modules/imgproc/src/contours_approx.cpp:
```diff
@@ -117,33 +122,36 @@ static int calc_cosine(const vector<ApproxItem> & ares, size_t i)
-        if( j < k && sk.i <= s )
+        temp_num = (float)(temp_num
+                           / sqrt(((double)dx1 * dx1 + (double)dy1 * dy1)
+                                  * ((double)dx2 * dx2 + (double)dy2 * dy2)));
```
This is fixed by setting `BreakBeforeBinaryOperators: None`

sn4k3 commented:
Would also be nice to have 4 connected and 8 connected contour detection...

vrabaud commented on modules/imgproc/src/contours_link.cpp:
```diff
@@ -0,0 +1,417 @@
+            lower_run = rns[rns[lower_run].next].next;
+            continue;
+        }
+        rns[rns[lower_run].next] = rns[rns[lower_run].next];
```
@mshabunin , why assign it to itself? Is it a bug?

mshabunin replied:
Yes, this is a problem, but not serious - there should be no operation, like in the original code:

https://github.com/opencv/opencv/blob/f4ebf7c0a6b2ee845a395ae3dac59927eb862894/modules/imgproc/src/contours.cpp#L1636-L1664

I'll fix it soon.
